File: src/physics/AABB.cpp
#include "poorcraft/physics/AABB.h"

#include <algorithm>

namespace PoorCraft {

PhysicsAABB::PhysicsAABB(const glm::vec3& minCorner, const glm::vec3& maxCorner)
    : m_Min(minCorner), m_Max(maxCorner) {
}

PhysicsAABB PhysicsAABB::fromCenterExtents(const glm::vec3& center, const glm::vec3& extents) {
    return PhysicsAABB(center - extents, center + extents);
}

glm::vec3 PhysicsAABB::getCenter() const {
    return (m_Min + m_Max) * 0.5f;
}

glm::vec3 PhysicsAABB::getExtents() const {
    return (m_Max - m_Min) * 0.5f;
}

const glm::vec3& PhysicsAABB::getMin() const {
    return m_Min;
}

const glm::vec3& PhysicsAABB::getMax() const {
    return m_Max;
}

bool PhysicsAABB::intersects(const PhysicsAABB& other) const {
    return (m_Min.x <= other.m_Max.x && m_Max.x >= other.m_Min.x) &&
           (m_Min.y <= other.m_Max.y && m_Max.y >= other.m_Min.y) &&
           (m_Min.z <= other.m_Max.z && m_Max.z >= other.m_Min.z);
}

bool PhysicsAABB::contains(const glm::vec3& point) const {
    return (point.x >= m_Min.x && point.x <= m_Max.x) &&
           (point.y >= m_Min.y && point.y <= m_Max.y) &&
           (point.z >= m_Min.z && point.z <= m_Max.z);
}

void PhysicsAABB::expand(float amount) {
    m_Min -= glm::vec3(amount);
    m_Max += glm::vec3(amount);
}

void PhysicsAABB::translate(const glm::vec3& offset) {
    m_Min += offset;
    m_Max += offset;
}

std::array<glm::vec3, 8> PhysicsAABB::getCorners() const {
    return {
        glm::vec3(m_Min.x, m_Min.y, m_Min.z),
        glm::vec3(m_Max.x, m_Min.y, m_Min.z),
        glm::vec3(m_Min.x, m_Max.y, m_Min.z),
        glm::vec3(m_Max.x, m_Max.y, m_Min.z),
        glm::vec3(m_Min.x, m_Min.y, m_Max.z),
        glm::vec3(m_Max.x, m_Min.y, m_Max.z),
        glm::vec3(m_Min.x, m_Max.y, m_Max.z),
        glm::vec3(m_Max.x, m_Max.y, m_Max.z)};
}

PhysicsAABB PhysicsAABB::merge(const PhysicsAABB& other) const {
    glm::vec3 minCorner(std::min(m_Min.x, other.m_Min.x),
                        std::min(m_Min.y, other.m_Min.y),
                        std::min(m_Min.z, other.m_Min.z));
    glm::vec3 maxCorner(std::max(m_Max.x, other.m_Max.x),
                        std::max(m_Max.y, other.m_Max.y),
                        std::max(m_Max.z, other.m_Max.z));
    return PhysicsAABB(minCorner, maxCorner);
}

} // namespace PoorCraft

File: common/src/main/java/com/poorcraft/common/physics/AABB.java
package com.poorcraft.common.physics;

import org.joml.Vector3f;

/**
 * Axis-aligned bounding box used for collision detection.
 */
public class AABB {

    private float minX;
    private float minY;
    private float minZ;
    private float maxX;
    private float maxY;
    private float maxZ;

    public AABB(Vector3f min, Vector3f max) {
        this(min.x, min.y, min.z, max.x, max.y, max.z);
    }

    public AABB(float minX, float minY, float minZ, float maxX, float maxY, float maxZ) {
        this.minX = Math.min(minX, maxX);
        this.minY = Math.min(minY, maxY);
        this.minZ = Math.min(minZ, maxZ);
        this.maxX = Math.max(minX, maxX);
        this.maxY = Math.max(minY, maxY);
        this.maxZ = Math.max(minZ, maxZ);
    }

    public boolean intersects(AABB other) {
        return this.maxX > other.minX && this.minX < other.maxX
            && this.maxY > other.minY && this.minY < other.maxY
            && this.maxZ > other.minZ && this.minZ < other.maxZ;
    }

    public boolean contains(Vector3f point) {
        return point.x >= minX && point.x <= maxX
            && point.y >= minY && point.y <= maxY
            && point.z >= minZ && point.z <= maxZ;
    }

    public AABB expand(float amount) {
        minX -= amount;
        minY -= amount;
        minZ -= amount;
        maxX += amount;
        maxY += amount;
        maxZ += amount;
        return this;
    }

    public AABB offset(Vector3f delta) {
        return offset(delta.x, delta.y, delta.z);
    }

    public AABB offset(float dx, float dy, float dz) {
        minX += dx;
        minY += dy;
        minZ += dz;
        maxX += dx;
        maxY += dy;
        maxZ += dz;
        return this;
    }

    public AABB union(AABB other) {
        return new AABB(
            Math.min(this.minX, other.minX),
            Math.min(this.minY, other.minY),
            Math.min(this.minZ, other.minZ),
            Math.max(this.maxX, other.maxX),
            Math.max(this.maxY, other.maxY),
            Math.max(this.maxZ, other.maxZ)
        );
    }

    public Vector3f getCenter() {
        return new Vector3f(
            (minX + maxX) * 0.5f,
            (minY + maxY) * 0.5f,
            (minZ + maxZ) * 0.5f
        );
    }

    public Vector3f getSize() {
        return new Vector3f(
            maxX - minX,
            maxY - minY,
            maxZ - minZ
        );
    }

    public static AABB fromEntity(Vector3f position, float width, float height) {
        float halfWidth = width * 0.5f;
        return new AABB(
            position.x - halfWidth,
            position.y,
            position.z - halfWidth,
            position.x + halfWidth,
            position.y + height,
            position.z + halfWidth
        );
    }

    public float getMinX() {
        return minX;
    }

    public float getMinY() {
        return minY;
    }

    public float getMinZ() {
        return minZ;
    }

    public float getMaxX() {
        return maxX;
    }

    public float getMaxY() {
        return maxY;
    }

    public float getMaxZ() {
        return maxZ;
    }

    public void setMinX(float minX) {
        this.minX = minX;
    }

    public void setMinY(float minY) {
        this.minY = minY;
    }

    public void setMinZ(float minZ) {
        this.minZ = minZ;
    }

    public void setMaxX(float maxX) {
        this.maxX = maxX;
    }

    public void setMaxY(float maxY) {
        this.maxY = maxY;
    }

    public void setMaxZ(float maxZ) {
        this.maxZ = maxZ;
    }

    @Override
    public String toString() {
        return "AABB{" +
            "minX=" + minX +
            ", minY=" + minY +
            ", minZ=" + minZ +
            ", maxX=" + maxX +
            ", maxY=" + maxY +
            ", maxZ=" + maxZ +
            '}';
    }
}
